### PR TITLE
Dev Solution 2  

### PR DESCRIPTION
**Pull request by taraspiotr**
- double unet pipeline
- change imageio to cv2 in masks preparation

### CODE REVIEW

File: src/loaders.py [hunk 1]
```diff
@@ -88,10 +89,12 @@ def __getitem__(self, index):
 
         if self.y is not None:
             Mi = self.y[0][index]
+            max_class = from_pil(Mi)[0].max()
 
             if self.train_mode and self.image_augment_with_target is not None:
                 Xi, Mi = from_pil(Xi, Mi)
                 Xi, Mi = self.image_augment_with_target(Xi, Mi)
+                Mi = np.where(Mi > max_class, max_class, Mi)
```
Comment (taraspiotr) [inline src/loaders.py]: Another solution would be to drop augmentations with target.

Comment (jakubczakon) [reply src/loaders.py]: @taraspiotr let's sleep on this one. There should be a nicer way to do it

File: neptune.yaml [hunk 2]
```diff
@@ -52,13 +49,18 @@ parameters:
   image_channels: 3
 
 # U-Net parameters
-  nr_unet_outputs: 4
+  unet_output_channels: 2
```
Comment (jakubczakon) [inline neptune.yaml]: @taraspiotr is this for the solution-0 ?
Also I think that unet_mask_... is more descriptive than unet_simple... 
Can we drop one of those (unet...) and keep two configs like neptune_solution_0.yaml neptune_solution_1.yaml or going one step further have `cfgs` dir and in there have solution_0.yaml solution_1_masks.yaml solution_1_masks_with_borders.yaml ?

File: src/augmentation.py [hunk 3]
```diff
@@ -1,5 +1,32 @@
+import cv2
+import numpy as np
+import imgaug as ia
 from imgaug import augmenters as iaa
 
+
+class PerspectiveTransform(iaa.PerspectiveTransform):
```
Comment (jakubczakon) [inline src/augmentation.py]: @taraspiotr I am not sure if I like that we are inheriting from implementation class and the interface iaa.Augmenter class. There is a tiny bit of boiler-plating involved if we chose to do it by the book but I'd still go for it.

File: src/models.py [hunk 4]
```diff
@@ -46,14 +48,18 @@ def __init__(self, architecture_config, training_config, callbacks_config):
         self.weight_regularization = weight_regularization_unet
         self.optimizer = optim.Adam(self.weight_regularization(self.model, **architecture_config['regularizer_params']),
                                     **architecture_config['optimizer_params'])
-        self.loss_function = [('mask', multiclass_segmentation_loss, 1.0)]
+        dice_loss = partial(multiclass_dice_loss,
+                            excluded_classes=[0])
+        loss_function = partial(mixed_dice_cross_entropy_loss,
```
Comment (jakubczakon) [inline src/models.py]: @taraspiotr the `loss_function` is implemented as a list of tuples to be able to pass multiple losses (and weigh them). I know in mapping there were other things involved and for simplicity we went with one function. Nonetheless I think this is a perfect example where one can (and I believe should) use the designed interface for passing weights and losses

File: src/models.py [hunk 5]
```diff
@@ -119,3 +104,52 @@ def callbacks_unet(callbacks_config):
     return CallbackList(
         callbacks=[experiment_timing, training_monitor, validation_monitor,
                    model_checkpoints, lr_scheduler, neptune_monitor, early_stopping])
+
+
+def mixed_dice_cross_entropy_loss(output, target, dice_weight=0.5, dice_loss=None,
```
Comment (jakubczakon) [inline src/models.py]: @taraspiotr I think instead of creating this function it is beter to pass 2 functions to the self.loss_function list of tuples

File: src/pipelines.py [hunk 6]
```diff
@@ -40,61 +37,48 @@ def unet(config, train_mode):
...
-                    transformer=Postprocessor(),
-                    input_steps=[mask_resize, contour_resize],
-                    adapter={'images': ([(mask_resize.name, 'resized_images')]),
-                             'contours': ([(contour_resize.name, 'resized_images')]),
-                             },
-                    cache_dirpath=config.env.cache_dirpath,
-                    save_output=save_output)
+    borders = mask_postprocessing_borders(unet_borders, config, save_output=save_output)
+
+    watersheder = Step(name='watershed',
```
Comment (jakubczakon) [inline src/pipelines.py]: I love the name watersheder. Sounds like a superhero! 
"Is it a bird is it a morphological closing transform. No it's a watersheder"

File: src/pipelines.py [hunk 7]
```diff
@@ -345,117 +226,79 @@ def mask_postprocessing(model, config, save_output=False):
...
-                                                             output_name='resized_images',
-                                                             apply_on=['images', 'target_sizes']),
-                          input_data=['input'],
-                          input_steps=[model],
-                          adapter={'images': ([(model.name, 'contour_prediction')]),
-                                   'target_sizes': ([('input', 'target_sizes')]),
-                                   },
-                          cache_dirpath=config.env.cache_dirpath,
-                          save_output=save_output)
+def mask_postprocessing_simple(model, config, save_output=False):
```
Comment (jakubczakon) [inline src/pipelines.py]: @taraspiotr I don't like the `_simple` suffix (though I am culprit of using it too). It is to generic. Why not `_masks` ?

File: src/pipelines.py [hunk 8]
```diff
@@ -345,117 +226,79 @@ def mask_postprocessing(model, config, save_output=False):
...
-                                 },
-
-                        cache_dirpath=config.env.cache_dirpath,
-                        save_output=save_output)
-    return drop_smaller
+                           cache_dirpath=config.env.cache_dirpath,
+                           save_output=save_output)
 
+    borders_filter = Step(name='borders_filter_borders',
+                          transformer=make_apply_transformer(partial(get_channel,
```
Comment (jakubczakon) [inline src/pipelines.py]: @taraspiotr those `make_apply_transformer`'s make this code really nice and simple. It was a really good idea @taraspiotr

File: src/pipelines.py [hunk 9]
```diff
@@ -345,117 +226,79 @@ def mask_postprocessing(model, config, save_output=False):
...
-                   input_steps=[postprocessed_mask],
-                   adapter={'images': ([(postprocessed_mask.name, 'binarized_images')]),
-                            },
-                   cache_dirpath=config.env.cache_dirpath,
-                   save_output=save_output)
-    return labeler
+    return borders_filter
 
 
 PIPELINES = {'unet': {'train': partial(unet, train_mode=True),
```
Comment (jakubczakon) [inline src/pipelines.py]: @taraspiotr I think we should have unet_masks and unet_masks_with_borders pipelines for training (as I believe you suggested). It would be easier for people to reproduce it easily

File: neptune.yaml (hunk 2 above)
Comment (apyskir) [reply neptune.yaml]: @taraspiotr Please don't remove `nr_unet_outputs` parameter. It is needed when you build UNet from scratch to determine number of output tensors from UNet.

File: src/loaders.py [hunk 10]
```diff
@@ -100,109 +101,8 @@ def __getitem__(self, index):
...
-                data[0] = self.image_augment(data[0])
-                data = to_pil(*data)
-
-            if self.mask_transform is not None:
-                data[1:] = [self.mask_transform(mask) for mask in data[1:]]
-
-            if self.image_transform is not None:
-                data[0] = self.image_transform(data[0])
-
-            return data
```
Comment (apyskir) [inline src/loaders.py]: OK, so now we remove the multi-output UNet, right? @jakubczakon are you aware of that?

File: src/models.py (hunk 5 above)
Comment (apyskir) [reply src/models.py]: @jakubczakon As far as I remember, `self.loss` function expects as many elements in a list, as number of output legs from UNet. Here we have one leg, so one loss function is needed, right?

File: src/preparation.py [hunk 11]
```diff
@@ -6,21 +6,20 @@
...
+def train_valid_split(meta, validation_size):
     meta_train = meta[meta['is_train'] == 1]
-    meta_train_split, meta_valid_split = split_on_column(meta_train,
-                                                         column='vgg_features_clusters',
-                                                         test_size=validation_size,
-                                                         random_state=1234,
-                                                         valid_category_ids=valid_category_ids
-                                                         )
+
+    meta_train_split, meta_valid_split = train_test_split(meta_train, test_size=validation_size, random_state=1234)
```
Comment (apyskir) [inline src/preparation.py]: @taraspiotr I think we should have `random_state` parametrized in pipeline_config.py, as it was in Mapping Challenge.

File: src/models.py (hunk 5 above)
Comment (taraspiotr) [reply src/models.py]: @apyskir I was thinking the same thing

File: src/augmentation.py (hunk 3 above)
Comment (taraspiotr) [reply src/augmentation.py]: @jakubczakon inheriting from Augmenter would actually mean quite a lot of copy pasting, what do you think about this kind of monkey patch:
`iaa.PerspectiveTransform._augment_images = _perspective_transform_augment_images`?

File: src/models.py (hunk 4 above)
Comment (taraspiotr) [reply src/models.py]: @jakubczakon The same point as @apyskir, I believe multiple losses expects multiple outputs and targets, while we use only one multichannel output.

File: src/loaders.py (hunk 10 above)
Comment (taraspiotr) [reply src/loaders.py]: @apyskir I think we were talking about this, but waiting for solution 2 implementation with that. I could be wrong though.

File: src/augmentation.py (hunk 3 above)
Comment (jakubczakon) [reply src/augmentation.py]: @taraspiotr I don't have a strong opinion on monkey patching but I don't mind it here since this is exactly what we are doing -> patching buggish implementation

File: src/loaders.py (hunk 10 above)
Comment (jakubczakon) [reply src/loaders.py]: @taraspiotr @apyskir yes I am aware, remember that we are reimplementing the winning solution. Our solution still has this transformer. 

I will move those loader transformers to `steppy-toolkit` in the following weeks so we are not going to lose any of that.
@apyskir thanks for asking!

File: src/models.py (hunk 4 above)
Comment (jakubczakon) [reply src/models.py]: @taraspiotr correct. nvmd

File: src/models.py (hunk 5 above)
Comment (jakubczakon) [reply src/models.py]: @taraspiotr @apyskir correct. nvmd

File: src/preparation.py (hunk 11 above)
Comment (jakubczakon) [reply src/preparation.py]: @apyskir agree 100%

File: src/loaders.py (hunk 10 above)
Comment (apyskir) [reply src/loaders.py]: @jakubczakon @taraspiotr OK, got it.
